src/main/java/tigase/xmpp/XMPPIOService.java

```diff
@@ -125,6 +125,9 @@ public void addPacketToSend(Packet packet) {
 			log.log(Level.FINEST, "Added packet to send: {1} [{0}]", new Object[]{toString(), packet});
 		}
 
+		if (getAuthorisedUserJid().isPresent() && packet.getStanzaTo() == null) {
```

Comment:
I'm not sure about this. It will set `to` attribute and this will (most likely) break `<presence/>` broadcast as it is done only if `to` is `null`.

While I agree, we may set `from` I'm against setting `to` as this will cause issues with routing and matching responses on the clients (with changed `to` in request, the response will have different `from` than the client expected). Moreover, while I could potentially agree on setting `to` to bare JID,  you are setting it here to full jid, which I think is an error.

Reply:
> I'm not sure about this. It will set to attribute and this will (most likely) break <presence/> broadcast as it is done only if to is null.

Why? It's `tigase.xmpp.XMPPIOService#addPacketToSend()` method which is used to process and queue packets that are to be written to socket (at the end we add them to `waitingPackets.offer(packet);` thus it's one of the last steps before the packets is going to be written to socket and thus delivered to client, i.e. I don't see how it could affect in any way broadcasting of the presence. I agree with your comment regarding setting `to` but that's in the opposite direction.

Though, I see one serious issue in the above code - I initially though about putting it `tigase.server.xmppclient.StreamManagementIOProcessor#processOutgoing()` (main culprit here) but decided to "generalise" it and ended up in `XMPPIOService` which is used by all ConnectionManagers (and not only ClientConnectionManager) so this could impact ClusterConnectionManager. In the end I think it should be moved back to tigase.server.xmppclient.StreamManagementIOProcessor#processOutgoing()`

Reply:
Yes, we should move it to `StreamManagementIOProcessor#processOutgoing()`. Initially, I thought that this was for processing data read from the socket. If we are writing to it then most likely it is OK.